### PR TITLE
audio_stream: do shifts instead of divisions where possible

### PR DESCRIPTION
**Pull request by lgirdwood**
Optimise free/avail sample/frame APIs to shift instead of divide where possible.

### CODE REVIEW

**Comment by lgirdwood** [comment]
@andrula-song can you take this over and do a perf analysis. @mwasko @lyakh @singalsu @serhiy-katsyuba-intel fyi

**Comment by lyakh** [comment]
@lgirdwood @andresoportus please also have a look at #7731

**Comment by singalsu** [comment]
On TGL UPX, git main 6fb22d645ef673034efdcdc1dd7bc94197db5d59 is 20.6 MCPS, with this patch 20.6 MCPS, no improvement seen.

File: src/include/sof/audio/audio_stream.h [hunk 1]
```diff
@@ -515,8 +537,18 @@ audio_stream_get_free_samples(const struct audio_stream __sparse_cache *stream)
...
-		audio_stream_frame_bytes(stream);
+	uint32_t frame_bytes = audio_stream_frame_bytes(stream);
+
+	switch (frame_bytes) {
+	case 2:
+		return audio_stream_get_free_bytes(stream) >> 1;
+	case 4:
+		return audio_stream_get_free_bytes(stream) >> 2;
+	case 8:
+		return audio_stream_get_free_bytes(stream) >> 3;
```
Comment (btian1) [inline src/include/sof/audio/audio_stream.h:548]: there is no case for 8?

File: src/include/sof/audio/audio_stream.h [hunk 2]
```diff
@@ -515,8 +537,18 @@ audio_stream_get_free_samples(const struct audio_stream __sparse_cache *stream)
...
 {
-	return audio_stream_get_free_bytes(stream) /
-		audio_stream_frame_bytes(stream);
+	uint32_t frame_bytes = audio_stream_frame_bytes(stream);
+
+	switch (frame_bytes) {
+	case 2:
+		return audio_stream_get_free_bytes(stream) >> 1;
+	case 4:
+		return audio_stream_get_free_bytes(stream) >> 2;
```
Comment (btian1) [inline src/include/sof/audio/audio_stream.h:546]: we'd better try to save switch, simply:
audio_stream_get_free_bytes(stream) >> (frame_bytes >> 1)

**Comment by andrula-song** [comment]
> On TGL UPX, git main [6fb22d6](https://github.com/thesofproject/sof/commit/6fb22d645ef673034efdcdc1dd7bc94197db5d59) is 20.6 MCPS, with this patch 20.6 MCPS, no improvement seen.

did your test pipeline contain the mixin mixout? if not, the result is expected since almost all module use comp_get_copy_limits_frame_aligned to get the frame number to process.

**Comment by andrula-song** [comment]
here is the result's from xtensa simulator:
image
the optimized audio_stream_get_free_frames can save about 16% cycles;
the optimized audio_stream_get_free_samples can save about 14% cycles;
the optimized audio_stream_get_avail_frames can save about 16.6% cycles;
the optimized audio_stream_get_avail_samples can save about 12% cycles;

**Comment by btian1** [comment]
However, on machine, according to below result, there is no saving for performance, check in #7766 

Before:

host-copier.0.playback avg 3.3 max 6.6
gain.1.1 avg 2.3 max 9.2
mixin.1.1 avg 2.7 max 3.0
mixout.2.1 avg 2.2 max 2.7
aria.2.1 avg 5.2 max 5.6
gain.2.1 avg 2.4 max 16.2
dai-copier.HDA.Analog.playback avg 4.1 max 4.3

Sum of average MCPS 22.3

After:

host-copier.0.playback avg 3.4 max 6.5
gain.1.1 avg 2.3 max 9.3
mixin.1.1 avg 2.7 max 3.0
mixout.2.1 avg 2.2 max 2.7
aria.2.1 avg 5.3 max 5.5
gain.2.1 avg 2.4 max 17.1
dai-copier.HDA.Analog.playback avg 4.1 max 11.2

Sum of average MCPS 22.4

So no performance saving, closing this PR.

**Pull request closed without merging**